File: apps/go_backend/internal/features/categories/handler.go
package categories

import (
	"net/http"

	"github.com/gin-gonic/gin"

	"github.com/lucid-logs/go-backend/internal/shared/errors"
	"github.com/lucid-logs/go-backend/internal/shared/middleware"
	"github.com/lucid-logs/go-backend/internal/shared/pagination"
	"github.com/lucid-logs/go-backend/internal/shared/response"
	"github.com/lucid-logs/go-backend/internal/shared/validator"
)

// =============================================================================
// HANDLER
// =============================================================================

// Handler handles category HTTP endpoints.
type Handler struct {
	service   Service
	validator *validator.Validator
}

// NewHandler creates a new category Handler.
func NewHandler(service Service, validator *validator.Validator) *Handler {
	return &Handler{
		service:   service,
		validator: validator,
	}
}

// =============================================================================
// ROUTES
// =============================================================================

// RegisterRoutes registers the category routes.
func RegisterRoutes(r *gin.RouterGroup, service Service, validator *validator.Validator) {
	h := NewHandler(service, validator)

	r.GET("", h.List)
	r.POST("", h.Create)
	r.GET("/:id", h.Get)
	r.PUT("/:id", h.Update)
	r.DELETE("/:id", h.Delete)
}

// =============================================================================
// HANDLERS
// =============================================================================

// List handles GET /categories
//
// @Summary      List categories
// @Description  Get paginated categories for the authenticated user
// @Tags         categories
// @Produce      json
// @Param        limit  query int false "Items per page (default 20, max 100)"
// @Param        offset query int false "Items to skip (default 0)"
// @Param        search query string false "Search by category name"
// @Success      200 {object} categories.CategoryPageResponse
// @Failure      401 {object} response.APIResponse
// @Failure      500 {object} response.APIResponse
// @Security     BearerAuth
// @Router       /api/v1/categories [get]
func (h *Handler) List(c *gin.Context) {
	user, appErr := middleware.MustGetAuthenticatedUser(c.Request.Context())
	if appErr != nil {
		response.Error(c, appErr)
		return
	}

	params := pagination.FromRequest(c)
	search := c.Query("search")

	resp, err := h.service.List(c.Request.Context(), user.UserID, params, search)
	if err != nil {
		response.ErrorFromErr(c, err)
		return
	}

	response.OK(c, resp)
}

// Get handles GET /categories/{id}
//
// @Summary      Get category
// @Description  Get a single category by ID
// @Tags         categories
// @Produce      json
// @Param        id path string true "Category ID"
// @Success      200 {object} Category
// @Failure      401 {object} response.APIResponse
// @Failure      404 {object} response.APIResponse
// @Failure      500 {object} response.APIResponse
// @Security     BearerAuth
// @Router       /api/v1/categories/{id} [get]
func (h *Handler) Get(c *gin.Context) {
	user, appErr := middleware.MustGetAuthenticatedUser(c.Request.Context())
	if appErr != nil {
		response.Error(c, appErr)
		return
	}

	categoryID := c.Param("id")

	cat, err := h.service.Get(c.Request.Context(), categoryID, user.UserID)
	if err != nil {
		if errors.Is(err, errors.ErrNotFound) {
			response.NotFound(c)
			return
		}
		response.ErrorFromErr(c, err)
		return
	}

	response.OK(c, cat)
}

// Create handles POST /categories
//
// @Summary      Create category
// @Description  Create a category for organizing tasks
// @Tags         categories
// @Accept       json
// @Produce      json
// @Param        request body CreateRequest true "Category data"
// @Success      201 {object} Category
// @Failure      400 {object} response.APIResponse
// @Failure      401 {object} response.APIResponse
// @Failure      409 {object} response.APIResponse
// @Failure      500 {object} response.APIResponse
// @Security     BearerAuth
// @Router       /api/v1/categories [post]
func (h *Handler) Create(c *gin.Context) {
	user, appErr := middleware.MustGetAuthenticatedUser(c.Request.Context())
	if appErr != nil {
		response.Error(c, appErr)
		return
	}

	var req CreateRequest
	if err := c.ShouldBindJSON(&req); err != nil {
		response.BadRequest(c, "Invalid JSON body")
		return
	}

	if errs := h.validator.Validate(&req); errs != nil {
		response.ValidationFailed(c, errs)
		return
	}

	cat, err := h.service.Create(c.Request.Context(), &req, user.UserID)
	if err != nil {
		if errors.Is(err, errors.ErrCategoryNameExists) {
			response.Conflict(c, "Category with this name already exists")
			return
		}
		response.ErrorFromErr(c, err)
		return
	}

	response.Created(c, cat)
}

// Update handles PUT /categories/{id}
//
// @Summary      Update category
// @Description  Update the name or color of a category
// @Tags         categories
// @Accept       json
// @Produce      json
// @Param        id      path string        true "Category ID"
// @Param        request body UpdateRequest true "Update payload"
// @Success      200 {object} Category
// @Failure      400 {object} response.APIResponse
// @Failure      401 {object} response.APIResponse
// @Failure      404 {object} response.APIResponse
// @Failure      409 {object} response.APIResponse
// @Failure      500 {object} response.APIResponse
// @Security     BearerAuth
// @Router       /api/v1/categories/{id} [put]
func (h *Handler) Update(c *gin.Context) {
	user, appErr := middleware.MustGetAuthenticatedUser(c.Request.Context())
	if appErr != nil {
		response.Error(c, appErr)
		return
	}

	categoryID := c.Param("id")

	var req UpdateRequest
	if err := c.ShouldBindJSON(&req); err != nil {
		response.BadRequest(c, "Invalid JSON body")
		return
	}

	if errs := h.validator.Validate(&req); errs != nil {
		response.ValidationFailed(c, errs)
		return
	}

	cat, err := h.service.Update(c.Request.Context(), categoryID, &req, user.UserID)
	if err != nil {
		if errors.Is(err, errors.ErrNotFound) {
			response.NotFound(c)
			return
		}
		if errors.Is(err, errors.ErrCategoryNameExists) {
			response.Conflict(c, "Category with this name already exists")
			return
		}
		response.ErrorFromErr(c, err)
		return
	}

	response.OK(c, cat)
}

// Delete handles DELETE /categories/{id}
//
// @Summary      Delete category
// @Description  Soft delete a category
// @Tags         categories
// @Produce      json
// @Param        id path string true "Category ID"
// @Success      200 {object} response.OperationMessage
// @Failure      401 {object} response.APIResponse
// @Failure      404 {object} response.APIResponse
// @Failure      500 {object} response.APIResponse
// @Security     BearerAuth
// @Router       /api/v1/categories/{id} [delete]
func (h *Handler) Delete(c *gin.Context) {
	user, appErr := middleware.MustGetAuthenticatedUser(c.Request.Context())
	if appErr != nil {
		response.Error(c, appErr)
		return
	}

	categoryID := c.Param("id")

	err := h.service.Delete(c.Request.Context(), categoryID, user.UserID)
	if err != nil {
		if errors.Is(err, errors.ErrNotFound) {
			response.NotFound(c)
			return
		}
		response.ErrorFromErr(c, err)
		return
	}

	response.Message(c, http.StatusOK, "Category deleted")
}

File: apps/rust_backend/src/features/categories/handler.rs
//! Category HTTP handlers

use axum::{
    extract::{Path, Query, State},
    http::StatusCode,
    middleware,
    routing::{delete, get, post, put},
    Json, Router,
};
use validator::Validate;

use super::model::{Category, CreateCategoryRequest, UpdateCategoryRequest};
use crate::core::error::{ApiResponse, AppError, PaginatedResponse};
use crate::core::middleware::{auth_middleware, AuthenticatedUser};
use crate::shared::pagination::PaginationParams;
use crate::state::AppState;

/// Create category routes (protected by auth middleware)
pub fn routes() -> Router<AppState> {
    Router::new()
        .route("/categories", get(list_categories))
        .route("/categories", post(create_category))
        .route("/categories/{id}", get(get_category))
        .route("/categories/{id}", put(update_category))
        .route("/categories/{id}", delete(delete_category))
}

/// Create protected category routes with auth middleware applied
pub fn protected_routes(state: AppState) -> Router<AppState> {
    routes().layer(middleware::from_fn_with_state(state, auth_middleware))
}

/// List all categories with pagination
#[utoipa::path(
    get,
    path = "/api/v1/categories",
    params(PaginationParams),
    responses(
        (status = 200, description = "List of categories with pagination metadata"),
        (status = 400, description = "Invalid pagination parameters"),
        (status = 401, description = "Unauthorized"),
        (status = 500, description = "Internal server error")
    ),
    security(("bearer_auth" = [])),
    tag = "categories"
)]
pub async fn list_categories(
    State(state): State<AppState>,
    auth_user: AuthenticatedUser,
    Query(params): Query<PaginationParams>,
) -> Result<(StatusCode, Json<ApiResponse<PaginatedResponse<Category>>>), AppError> {
    params.validate()?;

    tracing::debug!(
        user_id = %auth_user.user_id,
        limit = %params.limit,
        offset = %params.offset,
        "listing categories"
    );

    let (categories, total) = state
        .category_service
        .list_categories(&auth_user.user_id, params.limit, params.offset)
        .await?;

    tracing::info!(count = %categories.len(), total = %total, "categories listed successfully");

    let response = PaginatedResponse::new(categories, total, params.limit, params.offset);
    Ok((StatusCode::OK, Json(ApiResponse::success(response))))
}

/// Create a new category
#[utoipa::path(
    post,
    path = "/api/v1/categories",
    request_body = CreateCategoryRequest,
    responses(
        (status = 201, description = "Category created successfully", body = Category),
        (status = 400, description = "Invalid request or duplicate name"),
        (status = 401, description = "Unauthorized"),
        (status = 500, description = "Internal server error")
    ),
    security(("bearer_auth" = [])),
    tag = "categories"
)]
pub async fn create_category(
    State(state): State<AppState>,
    auth_user: AuthenticatedUser,
    Json(req): Json<CreateCategoryRequest>,
) -> Result<(StatusCode, Json<ApiResponse<Category>>), AppError> {
    req.validate()?;

    tracing::debug!(
        user_id = %auth_user.user_id,
        name = %req.name,
        "creating category"
    );

    let category = state
        .category_service
        .create_category(req, &auth_user.user_id)
        .await?;

    tracing::info!(
        category_id = ?category.id,
        "category created successfully"
    );

    Ok((StatusCode::CREATED, Json(ApiResponse::success(category))))
}

/// Get a category by ID
#[utoipa::path(
    get,
    path = "/api/v1/categories/{id}",
    params(
        ("id" = String, Path, description = "Category ID")
    ),
    responses(
        (status = 200, description = "Category found", body = Category),
        (status = 401, description = "Unauthorized"),
        (status = 404, description = "Category not found"),
        (status = 500, description = "Internal server error")
    ),
    security(("bearer_auth" = [])),
    tag = "categories"
)]
pub async fn get_category(
    State(state): State<AppState>,
    auth_user: AuthenticatedUser,
    Path(id): Path<String>,
) -> Result<(StatusCode, Json<ApiResponse<Category>>), AppError> {
    tracing::debug!(
        category_id = %id,
        user_id = %auth_user.user_id,
        "getting category"
    );

    let category = state
        .category_service
        .get_category(&id, &auth_user.user_id)
        .await?;

    Ok((StatusCode::OK, Json(ApiResponse::success(category))))
}

/// Update a category
#[utoipa::path(
    put,
    path = "/api/v1/categories/{id}",
    params(
        ("id" = String, Path, description = "Category ID")
    ),
    request_body = UpdateCategoryRequest,
    responses(
        (status = 200, description = "Category updated successfully", body = Category),
        (status = 400, description = "Invalid request or duplicate name"),
        (status = 401, description = "Unauthorized"),
        (status = 404, description = "Category not found"),
        (status = 500, description = "Internal server error")
    ),
    security(("bearer_auth" = [])),
    tag = "categories"
)]
pub async fn update_category(
    State(state): State<AppState>,
    auth_user: AuthenticatedUser,
    Path(id): Path<String>,
    Json(req): Json<UpdateCategoryRequest>,
) -> Result<(StatusCode, Json<ApiResponse<Category>>), AppError> {
    req.validate()?;

    tracing::debug!(
        category_id = %id,
        user_id = %auth_user.user_id,
        "updating category"
    );

    let category = state
        .category_service
        .update_category(&id, req, &auth_user.user_id)
        .await?;

    tracing::info!(category_id = %id, "category updated successfully");

    Ok((StatusCode::OK, Json(ApiResponse::success(category))))
}

/// Delete a category (soft delete)
#[utoipa::path(
    delete,
    path = "/api/v1/categories/{id}",
    params(
        ("id" = String, Path, description = "Category ID")
    ),
    responses(
        (status = 204, description = "Category deleted successfully"),
        (status = 401, description = "Unauthorized"),
        (status = 404, description = "Category not found"),
        (status = 500, description = "Internal server error")
    ),
    security(("bearer_auth" = [])),
    tag = "categories"
)]
pub async fn delete_category(
    State(state): State<AppState>,
    auth_user: AuthenticatedUser,
    Path(id): Path<String>,
) -> Result<StatusCode, AppError> {
    tracing::debug!(
        category_id = %id,
        user_id = %auth_user.user_id,
        "deleting category"
    );

    state
        .category_service
        .delete_category(&id, &auth_user.user_id)
        .await?;

    tracing::info!(category_id = %id, "category deleted successfully");

    Ok(StatusCode::NO_CONTENT)
}
